pkg/auth/audit/audit.go

```diff
@@ -148,31 +166,74 @@ func (a *auditLog) write(userInfo *User, reqHeaders, resHeaders http.Header, res
 	}
 
 	var buffer bytes.Buffer
+
 	alByte, err := json.Marshal(a.log)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to marshal log message: %w", err)
 	}
 
 	buffer.Write(bytes.TrimSuffix(alByte, []byte("}")))
-	if a.writer.Level >= levelRequest && len(a.reqBody) > 0 {
-		buffer.WriteString(`,"requestBody":`)
-		buffer.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, a.reqBody), []byte("\n")))
-	}
-	if a.writer.Level >= levelRequestResponse && resHeaders.Get("Content-Type") == contentTypeJSON && len(resBody) > 0 {
-		buffer.WriteString(`,"responseBody":`)
-		buffer.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, resBody), []byte("\n")))
+	a.writeRequest(&buffer)
+
+	if err = a.writeResponse(&buffer, resHeaders, resBody); err != nil {
+		return err
 	}
+
 	buffer.WriteString("}")
 
 	var compactBuffer bytes.Buffer
 	err = json.Compact(&compactBuffer, buffer.Bytes())
 	if err != nil {
-		return errors.Wrap(err, "compact audit log json failed")
+		return fmt.Errorf("compact audit log json failed %w", err)
 	}
 
 	compactBuffer.WriteString("\n")
+
 	_, err = a.writer.Output.Write(compactBuffer.Bytes())
-	return err
+	if err != nil {
+		return fmt.Errorf("failed to write log to output: %w", err)
+	}
+
+	return nil
+}
+
+// writeRequest attempts to write the API request to the log message.
+func (a *auditLog) writeRequest(buf *bytes.Buffer) {
+	if a.writer.Level < LevelRequest || len(a.reqBody) == 0 {
+		return
+	}
+
+	buf.WriteString(`,"requestBody":`)
+	buf.Write(bytes.TrimSuffix(a.concealSensitiveData(a.log.RequestURI, a.reqBody), []byte("\n")))
+}
+
+// writeResponse attempt to write the API response to the log message.
+func (a *auditLog) writeResponse(buf *bytes.Buffer, resHeaders http.Header, resBody []byte) (err error) {
+	if a.writer.Level < LevelRequestResponse || resHeaders.Get("Content-Type") != contentTypeJSON || len(resBody) == 0 {
+		return nil
+	}
+
+	switch resHeaders.Get("Content-Encoding") {
```

Comment:
After going through our rancher code, apiserve handles the `content-encoding` and only uses `gzip`. We use a zip when returning key files for nodes, but this does not have the `Content-Type` of `application/json` so we do not need to handle the response. Zip would also be odd to handle since it is used for archiving multiple files.